cmd/maya-apiserver/cstor-operator/cspc/handler.go

```diff
@@ -249,88 +257,113 @@ func (pc *PoolConfig) createDeployForCSP(csp *apis.NewTestCStorPool) error {
+			return errors.Wrapf(err, "failed to handle CSPC %s deletion", pc.AlgorithmConfig.CSPC.Name)
 		}
 	}
 
-	// remove finalizer on cspc
-	err = c.removeCSPCFinalizer(cspc, v1alpha1.CSPCFinalizer)
-	if err != nil {
-		return errors.Wrapf(err, "failed to remove finalizer on cspc object")
-	}
-
-	// if finalizer is removed successfully, then object will
-	// be removed and no need to reconcile further
 	return nil
 }
 
-// removeBDCFinalizer removes the given finalizer from the BDC
-func (c *Controller) removeBDCFinalizer(bdcObj *ndmapis.BlockDeviceClaim, finalizer string) error {
-	if len(bdcObj.Finalizers) == 0 {
+// deleteAssociatedCSP deletes the CSP resource(s) belonging to the given CSPC resource.
+// If no CSP resource exists for the CSPC, then a levelled info log is logged and function
+// returns.
+func (pc *PoolConfig) deleteAssociatedCSP() error {
+	err := apiscsp.NewKubeClient().WithNamespace(pc.AlgorithmConfig.Namespace).DeleteCollection(
+		metav1.ListOptions{
+			LabelSelector: string(apis.CStorPoolClusterCPK) + "=" + pc.AlgorithmConfig.CSPC.Name,
+		},
+		&metav1.DeleteOptions{},
+	)
+
+	if k8serror.IsNotFound(err) {
+		glog.V(2).Infof("Associated CSP(s) of CSPC %s is already deleted:%s", pc.AlgorithmConfig.CSPC.Name, err.Error())
 		return nil
 	}
 
-	bdcObj.Finalizers = util.RemoveString(bdcObj.Finalizers, finalizer)
-
-	// Update is used instead of patch, because when there were 2 finalizers in the object
-	// and tried to remove the first finalizer using patch operation , it was not working. The
-	// patch operation didn't return any error but the object was not getting patched.
-	// using Update() it was possible to remove the finalizer on the BDC
-	_, err := blockdeviceclaim.NewKubeClient().
-		WithNamespace(bdcObj.Namespace).
-		Update(bdcObj)
 	if err != nil {
-		return errors.Wrapf(err, "failed to remove finalizer from BDC %v", bdcObj.Name)
+		return errors.Wrapf(err, "failed to delete associated CSP(s):%s", err.Error())
 	}
+	glog.Infof("Associated CSP(s) of CSPC %s deleted successfully ", pc.AlgorithmConfig.CSPC.Name)
 	return nil
 }
 
-// removeCSPCFinalizer will remove finalizer on cspc
-func (c *Controller) removeCSPCFinalizer(cspcObj *apis.CStorPoolCluster, finalizer string) error {
-	if len(cspcObj.Finalizers) == 0 {
-		return nil
+// removeSPCFinalizer removes CSPC finalizers on associated
+// BDC resources and CSPC object itself.
+func (pc *PoolConfig) removeCSPCFinalizer() error {
+	cspList, err := apiscsp.NewKubeClient().List(metav1.ListOptions{
+		LabelSelector: string(apis.StoragePoolClaimCPK) + "=" + pc.AlgorithmConfig.CSPC.Name,
+	})
+
+	if err != nil {
+		return errors.Wrap(err, "failed to remove CSPC finalizer on associated resources")
+	}
+
+	if len(cspList.Items) > 0 {
+		return errors.Wrap(err, "failed to remove CSPC finalizer on associated resources as "+
+			"CSP(s) still exists for CSPC")
+	}
+
+	err = pc.removeSPCFinalizerOnAssociatedBDC()
```

Comment:
let us make sure cstor-pool-mgmt pods are not running before removing finalizers on BDC